File: backend/app/schemas/leaderboard.py
"""
Leaderboard Schemas - Request/Response models
"""
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel
from uuid import UUID


class LeaderboardEntry(BaseModel):
    """Single entry in leaderboard"""
    rank: int
    user_id: UUID
    email: str
    full_name: Optional[str]
    roles: List[str]
    primary_role: str
    level: int
    current_xp: int
    discipline_score: float
    core_task_progress: float
    completed_tasks: int
    is_ready_to_promote: bool
    
    class Config:
        from_attributes = True


class LeaderboardResponse(BaseModel):
    """Leaderboard rankings"""
    entries: List[LeaderboardEntry]
    total_users: int
    my_rank: Optional[int] = None
    updated_at: datetime


class UserRankInfo(BaseModel):
    """User's personal ranking info"""
    rank: int
    total_users: int
    percentile: float
    current_xp: int
    next_rank_xp: Optional[int] = None
    xp_to_next_rank: Optional[int] = None

File: frontend/src/types/leaderboard.ts
/**
 * Leaderboard types
 */

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  email: string;
  full_name?: string | null;
  roles: string[];
  primary_role: string;
  level: number;
  current_xp: number;
  discipline_score: number;
  core_task_progress: number;
  completed_tasks: number;
  is_ready_to_promote: boolean;
}

export interface LeaderboardResponse {
  entries: LeaderboardEntry[];
  total_users: number;
  my_rank?: number | null;
  updated_at: string;
}

export interface UserRankInfo {
  rank: number;
  total_users: number;
  percentile: number;
  current_xp: number;
  next_rank_xp?: number | null;
  xp_to_next_rank?: number | null;
}

export interface LeaderboardStats {
  total_active_users: number;
  average_xp: number;
  average_level: number;
  highest_xp: number;
  ready_for_promotion: number;
}
